### ISSUE TITLE
Unpickleable=False Objects are null

### ISSUE THREAD

**Issue by tobias-huermann** [jsonpickle/jsonpickle#337]
I try to serialize a Python object tree, where some objects are the same. I want to avoid the py/id pointers (and all other bloat) from jsonpickle (1.4.2) so I use unpickleable=False. When i do it i get null values in the elements already present in the result JSON. Is it intended or a bug? I would suspect redundant data (deep-copy).

```python
import jsonpickle


class Person:
    def __init__(self):
        self.name = "Jane Doe"


class Car:
    def __init__(self):
        self.owner = None
        self.driver = None
        self.passengers = []


def test_jsonpickle():
    han = Person()
    han.name = "Han"
    car = Car()
    car.driver = han
    car.owner = han

    pa = Person()
    pa.name = "Pete"
    car.passengers.append(pa)
    car.passengers.append(han)

    json_string = jsonpickle.encode(car, unpicklable=False)
    print("Json String: " + json_string)
    # out: Json String: 
    # {"owner": {"name": "Han"}, "driver": null, "passengers": [{"name": "Pete"}, null]}


if __name__ == '__main__':
    test_jsonpickle()
```

Side note: 
I have a big class structure and the object trees need to be passed from python to JS context in JSON. I tried out jsonpickleJS and going with the py/id thing. Disadvantage is that i need to take care python and JS code bases are in synch as you need to register the available class names. Otherwise you end up with dangling pointers... 
As a workaround i have my own pointer implementation by making use of python `id()` memory addresses. These i resolve on my own on the JS side. Don't know exactly but wouldn't it be pretty neat if `jsonpickle` supports pointers / references by making use of the $id $ref features from the JSON spec.? But i am no expert there. I dont understand if thats only specified for JSON schema definitions or if it is intended to be used with actual data as well.

**Comment by davvid** [comment]
Interesting!  To be honest, the `unpicklable=False` code path has always been an after-thought, where things like repeated references would be dropped and it was fine/intentional.  I'd very much welcome improvements to that side of the `unpickable` boolean where it would do things like repeat the data, especially if we could cache the object dicts.

There aren't any cycles in the data, so it seems like we should be able to do something sensible.

I'm very much open to changing the format for `unpickable=False` if improves general usage, which this example demonstrates well.  I'm curious about the pointer implementation you mention.  If it's something where it caches the simple dict representation of objects, then that sounds generally useful and perhaps we can use it?

**Comment by davvid** [comment]
Please test the proposed changes in #338 and let me know if it fixes things for your use case.

My local testing shows that we are now getting:

```
{"driver": {"name": "Han"}, "owner": {"name": "Han"}, "passengers": [{"name": "Pete"}, {"name": "Han"}]}
```

...which looks like exactly what you were describing 😸

**Comment by davvid** [comment]
Fixed in jsonpickle v1.5.0 (just released today to pypi).  cheers

**Comment by tobias-huermann** [comment]
Awesome, thank you very much :) 
You asked for my pointer workaround. It does look like this:
```python
import jsonpickle

class Base:

    def __repr__(self):
        return self.get_object_id()

    def __getstate__(self):
        my_dict = self.__dict__
        my_dict["object_id"] = self.get_object_id()
        return my_dict

    def get_object_id(self):
        return hex(id(self))


class Person(Base):
    def __init__(self):
        self.name = "Jane Doe"


class Car(Base):
    def __init__(self):
        self.owner = None
        self.driver = None
        self.passengers = []


def test_jsonpickle():
    han = Person()
    han.name = "Han"
    car = Car()
    car.driver = han
    car.owner = han

    pa = Person()
    pa.name = "Pete"
    car.passengers.append(pa)
    car.passengers.append(han)

    json_string = jsonpickle.encode(car, make_refs=False, unpicklable=False, indent=2)
    print("Json String: " + json_string)

if __name__ == '__main__':
    test_jsonpickle()
```

So i didnt monkey patch and just introduced this base class for all data types. I get the following JSON:

```
Json String: {
  "owner": {
    "name": "Han",
    "object_id": "0x208fa045640"
  },
  "driver": "0x208fa045640",
  "passengers": [
    {
      "name": "Pete",
      "object_id": "0x208fbe68520"
    },
    "0x208fa045640"
  ],
  "object_id": "0x208fa0455e0"
}
```
This is quite handy and i can resolve the references in other clients quite easy 👍
